File: packages/nstreamdown/platforms/ios/src/SyntaxHighlighter.swift
import UIKit
import Foundation

/// Token types for syntax highlighting
@objc public enum SyntaxTokenType: Int {
    case keyword = 0
    case string = 1
    case number = 2
    case comment = 3
    case function = 4
    case type = 5
    case property = 6
    case punctuation = 7
    case plain = 8
}

/// Simple syntax highlighter exposed to NativeScript
/// Uses native NSAttributedString for efficient rendering
@objcMembers
public class SyntaxHighlighter: NSObject {
    
    // MARK: - Color Schemes
    
    public struct ColorScheme {
        let keyword: UIColor
        let string: UIColor
        let number: UIColor
        let comment: UIColor
        let function: UIColor
        let type: UIColor
        let property: UIColor
        let punctuation: UIColor
        let plain: UIColor
        let background: UIColor
    }
    
    public static let darkScheme = ColorScheme(
        keyword: UIColor(red: 0.75, green: 0.55, blue: 0.95, alpha: 1),     // purple
        string: UIColor(red: 0.30, green: 0.80, blue: 0.50, alpha: 1),      // green
        number: UIColor(red: 0.30, green: 0.82, blue: 0.88, alpha: 1),      // cyan
        comment: UIColor(red: 0.45, green: 0.50, blue: 0.56, alpha: 1),     // gray
        function: UIColor(red: 0.40, green: 0.65, blue: 0.95, alpha: 1),    // blue
        type: UIColor(red: 0.95, green: 0.80, blue: 0.30, alpha: 1),        // yellow
        property: UIColor(red: 0.90, green: 0.60, blue: 0.40, alpha: 1),    // orange
        punctuation: UIColor(red: 0.60, green: 0.65, blue: 0.72, alpha: 1), // light gray
        plain: UIColor(red: 0.90, green: 0.92, blue: 0.95, alpha: 1),       // white
        background: UIColor(red: 0.10, green: 0.12, blue: 0.16, alpha: 1)   // dark
    )
    
    // MARK: - Language Keywords
    
    private static let swiftKeywords: Set<String> = [
        "import", "class", "struct", "enum", "protocol", "extension", "func", "var", "let",
        "if", "else", "guard", "switch", "case", "default", "for", "while", "repeat", "do",
        "break", "continue", "return", "throw", "try", "catch", "defer", "in", "where",
        "as", "is", "nil", "true", "false", "self", "Self", "super", "init", "deinit",
        "get", "set", "willSet", "didSet", "throws", "rethrows", "async", "await",
        "override", "final", "static", "private", "public", "internal", "open", "fileprivate",
        "mutating", "nonmutating", "lazy", "weak", "unowned", "optional", "required",
        "convenience", "dynamic", "infix", "prefix", "postfix", "operator", "associatedtype",
        "typealias", "some", "any", "@objc", "@objcMembers", "@escaping", "@MainActor"
    ]
    
    private static let tsKeywords: Set<String> = [
        "const", "let", "var", "function", "class", "interface", "type", "enum",
        "if", "else", "for", "while", "do", "switch", "case", "default", "break", "continue",
        "return", "throw", "try", "catch", "finally", "new", "delete", "typeof", "instanceof",
        "import", "export", "from", "default", "as", "async", "await", "yield",
        "this", "super", "extends", "implements", "static", "get", "set",
        "public", "private", "protected", "readonly", "abstract", "declare",
        "null", "undefined", "true", "false", "void", "never", "any", "unknown",
        "string", "number", "boolean", "object", "symbol", "bigint"
    ]
    
    private static let pythonKeywords: Set<String> = [
        "def", "class", "if", "elif", "else", "for", "while", "try", "except", "finally",
        "with", "as", "import", "from", "return", "yield", "raise", "pass", "break", "continue",
        "and", "or", "not", "in", "is", "lambda", "global", "nonlocal", "assert", "del",
        "True", "False", "None", "async", "await", "self", "cls"
    ]
    
    private static let swiftTypes: Set<String> = [
        "Int", "String", "Bool", "Double", "Float", "Character", "Array", "Dictionary",
        "Set", "Optional", "Any", "AnyObject", "Void", "Never", "Result", "Error",
        "UIView", "UIViewController", "UILabel", "UIButton", "UIImage", "UIColor",
        "CGRect", "CGPoint", "CGSize", "CGFloat", "NSObject", "NSString", "NSArray"
    ]
    
    private static let tsTypes: Set<String> = [
        "Array", "Object", "String", "Number", "Boolean", "Function", "Promise",
        "Map", "Set", "WeakMap", "WeakSet", "Date", "RegExp", "Error", "Symbol",
        "Record", "Partial", "Required", "Readonly", "Pick", "Omit", "Exclude", "Extract",
        "Component", "Observable", "Subject", "EventEmitter"
    ]
    
    // MARK: - Shared Instance
    
    public static let shared = SyntaxHighlighter()
    
    private let monoFont: UIFont
    
    private override init() {
        // Use SF Mono or fallback to Menlo
        if let sfMono = UIFont(name: "SFMono-Regular", size: 13) {
            self.monoFont = sfMono
        } else {
            self.monoFont = UIFont.monospacedSystemFont(ofSize: 13, weight: .regular)
        }
        super.init()
    }
    
    // Background queue for highlighting
    private let highlightQueue = DispatchQueue(label: "com.streamdown.syntax", qos: .userInitiated)
    
    // MARK: - Public API (exposed to NativeScript)
    
    /// Highlight code and return an attributed string
    public func highlight(_ code: String, language: String) -> NSAttributedString {
        let scheme = Self.darkScheme
        let attributedString = NSMutableAttributedString(string: code)
        let fullRange = NSRange(location: 0, length: code.utf16.count)
        
        // Apply base attributes
        attributedString.addAttribute(.font, value: monoFont, range: fullRange)
        attributedString.addAttribute(.foregroundColor, value: scheme.plain, range: fullRange)
        
        // Get keywords and types for the language
        let keywords: Set<String>
        let types: Set<String>
        
        switch language.lowercased() {
        case "swift":
            keywords = Self.swiftKeywords
            types = Self.swiftTypes
        case "typescript", "ts", "javascript", "js":
            keywords = Self.tsKeywords
            types = Self.tsTypes
        case "python", "py":
            keywords = Self.pythonKeywords
            types = []
        default:
            keywords = Self.tsKeywords
            types = Self.tsTypes
        }
        
        // Tokenize and highlight
        highlightTokens(in: attributedString, code: code, keywords: keywords, types: types, scheme: scheme)
        
        return attributedString
    }
    
    /// Full highlight with tokenization (alias for highlight)
    public func highlightFull(_ code: String, language: String) -> NSAttributedString {
        return highlight(code, language: language)
    }
    
    /// Highlight code asynchronously (for NativeScript - uses callback)
    /// - Parameters:
    ///   - code: The source code to highlight
    ///   - language: The programming language
    ///   - label: The UILabel to update with the result
    public func highlightAsync(_ code: String, language: String, label: UILabel) {
        // Debug: Log immediately when method is called
        print("[SyntaxHighlighter] highlightAsync called with \(code.count) chars, language: \(language)")
        
        // Capture values for background thread
        let codeCopy = code
        let langCopy = language
        
        // Keep strong reference to label
        let labelRef = label
        
        print("[SyntaxHighlighter] About to dispatch to background queue...")
        
        highlightQueue.async { [weak self] in
            print("[SyntaxHighlighter] Inside background queue")
            guard let self = self else { 
                print("[SyntaxHighlighter] self is nil!")
                return 
            }
            
            // Do expensive highlighting on background thread
            print("[SyntaxHighlighter] Calling highlight()...")
            let attributedString = self.highlight(codeCopy, language: langCopy)
            print("[SyntaxHighlighter] Got attributed string with \(attributedString.length) chars")
            
            // Update UI on main thread
            DispatchQueue.main.async {
                print("[SyntaxHighlighter] On main thread, updating label...")
                labelRef.attributedText = attributedString
                print("[SyntaxHighlighter] Label updated!")
            }
        }
        
        print("[SyntaxHighlighter] highlightAsync returning")
    }
    
    /// Create a configured UITextView with highlighted code
    /// - Parameters:
    ///   - code: The source code to highlight
    ///   - language: The programming language
    ///   - frame: The frame for the text view
    /// - Returns: Configured UITextView
    public func createCodeView(_ code: String, language: String, frame: CGRect) -> UITextView {
        let textView = UITextView(frame: frame)
        textView.isEditable = false
        textView.isSelectable = true
        textView.backgroundColor = Self.darkScheme.background
        textView.textContainerInset = UIEdgeInsets(top: 12, left: 12, bottom: 12, right: 12)
        textView.attributedText = highlight(code, language: language)
        return textView
    }
    
    // MARK: - Tokenization (Fast regex-based)
    
    // Pre-compiled regex patterns for speed
    private static let stringPattern = try! NSRegularExpression(pattern: #"\"[^\"\\]*(?:\\.[^\"\\]*)*\"|'[^'\\]*(?:\\.[^'\\]*)*'|`[^`]*`"#, options: [])
    private static let commentPattern = try! NSRegularExpression(pattern: #"//[^\n]*|/\*[\s\S]*?\*/|#[^\n]*"#, options: [])
    private static let numberPattern = try! NSRegularExpression(pattern: #"\b\d+\.?\d*\b"#, options: [])
    private static let wordPattern = try! NSRegularExpression(pattern: #"\b[a-zA-Z_$][a-zA-Z0-9_$]*\b"#, options: [])
    
    private func highlightTokens(
        in attrString: NSMutableAttributedString,
        code: String,
        keywords: Set<String>,
        types: Set<String>,
        scheme: ColorScheme
    ) {
        let range = NSRange(location: 0, length: code.utf16.count)
        
        // Track which ranges are already colored (comments/strings take precedence)
        var coloredRanges = [NSRange]()
        
        // 1. Comments first (highest precedence)
        let comments = Self.commentPattern.matches(in: code, options: [], range: range)
        for match in comments {
            attrString.addAttribute(.foregroundColor, value: scheme.comment, range: match.range)
            coloredRanges.append(match.range)
        }
        
        // 2. Strings
        let strings = Self.stringPattern.matches(in: code, options: [], range: range)
        for match in strings {
            if !coloredRanges.contains(where: { NSIntersectionRange($0, match.range).length > 0 }) {
                attrString.addAttribute(.foregroundColor, value: scheme.string, range: match.range)
                coloredRanges.append(match.range)
            }
        }
        
        // 3. Numbers
        let numbers = Self.numberPattern.matches(in: code, options: [], range: range)
        for match in numbers {
            if !coloredRanges.contains(where: { NSIntersectionRange($0, match.range).length > 0 }) {
                attrString.addAttribute(.foregroundColor, value: scheme.number, range: match.range)
            }
        }
        
        // 4. Keywords and identifiers
        let codeNS = code as NSString
        let words = Self.wordPattern.matches(in: code, options: [], range: range)
        for match in words {
            if coloredRanges.contains(where: { NSIntersectionRange($0, match.range).length > 0 }) {
                continue
            }
            
            let word = codeNS.substring(with: match.range)
            
            if keywords.contains(word) {
                attrString.addAttribute(.foregroundColor, value: scheme.keyword, range: match.range)
            } else if types.contains(word) || (word.first?.isUppercase == true && word.count > 1) {
                attrString.addAttribute(.foregroundColor, value: scheme.type, range: match.range)
            } else {
                // Check if followed by ( for function
                let afterWord = match.range.location + match.range.length
                if afterWord < code.utf16.count {
                    var idx = afterWord
                    while idx < code.utf16.count {
                        let c = codeNS.character(at: idx)
                        if c == 32 || c == 9 { // space or tab
                            idx += 1
                            continue
                        }
                        if c == 40 { // (
                            attrString.addAttribute(.foregroundColor, value: scheme.function, range: match.range)
                        }
                        break
                    }
                }
            }
        }
    }
}

File: packages/nstreamdown/platforms/android/java/org/nativescript/streamdown/SyntaxHighlighter.kt
package org.nativescript.streamdown

import android.graphics.Color
import android.graphics.Typeface
import android.os.Handler
import android.os.Looper
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.BackgroundColorSpan
import android.text.style.ForegroundColorSpan
import android.text.style.StyleSpan
import android.text.style.TypefaceSpan
import android.widget.TextView
import java.util.concurrent.Executors
import java.util.regex.Pattern

/**
 * Token types for syntax highlighting
 */
enum class SyntaxTokenType {
    KEYWORD,
    STRING,
    NUMBER,
    COMMENT,
    FUNCTION,
    TYPE,
    PROPERTY,
    PUNCTUATION,
    PLAIN
}

/**
 * Color scheme for syntax highlighting
 */
data class ColorScheme(
    val keyword: Int,
    val string: Int,
    val number: Int,
    val comment: Int,
    val function: Int,
    val type: Int,
    val property: Int,
    val punctuation: Int,
    val plain: Int,
    val background: Int
)

/**
 * Syntax highlighter for NativeScript Streamdown
 * Uses Android SpannableString for efficient rendering
 */
class SyntaxHighlighter private constructor() {

    companion object {
        private val _instance: SyntaxHighlighter by lazy { SyntaxHighlighter() }
        
        @JvmStatic
        fun getShared(): SyntaxHighlighter = _instance

        // Dark color scheme (matching iOS VSCode-like colors)
        private val _darkScheme = ColorScheme(
            keyword = Color.rgb(86, 156, 214),      // #569CD6 - blue
            string = Color.rgb(206, 145, 120),      // #CE9178 - salmon/orange
            number = Color.rgb(181, 206, 168),      // #B5CEA8 - light green
            comment = Color.rgb(106, 153, 85),      // #6A9955 - green
            function = Color.rgb(220, 220, 170),    // #DCDCAA - yellow
            type = Color.rgb(78, 201, 176),         // #4EC9B0 - teal/cyan
            property = Color.rgb(156, 220, 254),    // #9CDCFE - light blue
            punctuation = Color.rgb(212, 212, 212), // #D4D4D4 - light gray
            plain = Color.rgb(212, 212, 212),       // #D4D4D4 - light gray
            background = Color.rgb(30, 30, 30)      // #1E1E1E - dark (VSCode dark)
        )

        @JvmStatic
        fun getDarkScheme(): ColorScheme = _darkScheme

        // Light color scheme (VSCode light-like colors)
        private val _lightScheme = ColorScheme(
            keyword = Color.rgb(0, 0, 255),         // #0000FF - blue
            string = Color.rgb(163, 21, 21),        // #A31515 - dark red
            number = Color.rgb(9, 136, 90),         // #09885A - green
            comment = Color.rgb(0, 128, 0),         // #008000 - green
            function = Color.rgb(121, 94, 38),      // #795E26 - brown
            type = Color.rgb(38, 127, 153),         // #267F99 - teal
            property = Color.rgb(0, 16, 128),       // #001080 - dark blue
            punctuation = Color.rgb(0, 0, 0),       // #000000 - black
            plain = Color.rgb(0, 0, 0),             // #000000 - black
            background = Color.rgb(255, 255, 255)   // #FFFFFF - white
        )

        @JvmStatic
        fun getLightScheme(): ColorScheme = _lightScheme

        // Language keywords
        private val swiftKeywords = setOf(
            "import", "class", "struct", "enum", "protocol", "extension", "func", "var", "let",
            "if", "else", "guard", "switch", "case", "default", "for", "while", "repeat", "do",
            "break", "continue", "return", "throw", "try", "catch", "defer", "in", "where",
            "as", "is", "nil", "true", "false", "self", "Self", "super", "init", "deinit",
            "get", "set", "willSet", "didSet", "throws", "rethrows", "async", "await",
            "override", "final", "static", "private", "public", "internal", "open", "fileprivate",
            "mutating", "nonmutating", "lazy", "weak", "unowned", "optional", "required",
            "convenience", "dynamic", "infix", "prefix", "postfix", "operator", "associatedtype",
            "typealias", "some", "any", "@objc", "@objcMembers", "@escaping", "@MainActor"
        )

        private val tsKeywords = setOf(
            "const", "let", "var", "function", "class", "interface", "type", "enum",
            "if", "else", "for", "while", "do", "switch", "case", "default", "break", "continue",
            "return", "throw", "try", "catch", "finally", "new", "delete", "typeof", "instanceof",
            "import", "export", "from", "default", "as", "async", "await", "yield",
            "this", "super", "extends", "implements", "static", "get", "set",
            "public", "private", "protected", "readonly", "abstract", "declare",
            "null", "undefined", "true", "false", "void", "never", "any", "unknown",
            "string", "number", "boolean", "object", "symbol", "bigint"
        )

        private val pythonKeywords = setOf(
            "def", "class", "if", "elif", "else", "for", "while", "try", "except", "finally",
            "with", "as", "import", "from", "return", "yield", "raise", "pass", "break", "continue",
            "and", "or", "not", "in", "is", "lambda", "global", "nonlocal", "assert", "del",
            "True", "False", "None", "async", "await", "self", "cls"
        )

        private val javaKeywords = setOf(
            "abstract", "assert", "boolean", "break", "byte", "case", "catch", "char", "class",
            "const", "continue", "default", "do", "double", "else", "enum", "extends", "final",
            "finally", "float", "for", "goto", "if", "implements", "import", "instanceof", "int",
            "interface", "long", "native", "new", "package", "private", "protected", "public",
            "return", "short", "static", "strictfp", "super", "switch", "synchronized", "this",
            "throw", "throws", "transient", "try", "void", "volatile", "while", "true", "false", "null"
        )

        private val kotlinKeywords = setOf(
            "abstract", "actual", "annotation", "as", "break", "by", "catch", "class", "companion",
            "const", "constructor", "continue", "crossinline", "data", "do", "dynamic", "else",
            "enum", "expect", "external", "false", "final", "finally", "for", "fun", "get", "if",
            "import", "in", "infix", "init", "inline", "inner", "interface", "internal", "is",
            "lateinit", "noinline", "null", "object", "open", "operator", "out", "override",
            "package", "private", "protected", "public", "reified", "return", "sealed", "set",
            "super", "suspend", "tailrec", "this", "throw", "true", "try", "typealias", "typeof",
            "val", "var", "vararg", "when", "where", "while"
        )

        private val goKeywords = setOf(
            "break", "case", "chan", "const", "continue", "default", "defer", "else", "fallthrough",
            "for", "func", "go", "goto", "if", "import", "interface", "map", "package", "range",
            "return", "select", "struct", "switch", "type", "var", "true", "false", "nil", "iota"
        )

        private val rustKeywords = setOf(
            "as", "async", "await", "break", "const", "continue", "crate", "dyn", "else", "enum",
            "extern", "false", "fn", "for", "if", "impl", "in", "let", "loop", "match", "mod",
            "move", "mut", "pub", "ref", "return", "self", "Self", "static", "struct", "super",
            "trait", "true", "type", "unsafe", "use", "where", "while"
        )

        private val cppKeywords = setOf(
            "alignas", "alignof", "and", "and_eq", "asm", "auto", "bitand", "bitor", "bool", "break",
            "case", "catch", "char", "char16_t", "char32_t", "class", "compl", "const", "constexpr",
            "const_cast", "continue", "decltype", "default", "delete", "do", "double", "dynamic_cast",
            "else", "enum", "explicit", "export", "extern", "false", "float", "for", "friend", "goto",
            "if", "inline", "int", "long", "mutable", "namespace", "new", "noexcept", "not", "not_eq",
            "nullptr", "operator", "or", "or_eq", "private", "protected", "public", "register",
            "reinterpret_cast", "return", "short", "signed", "sizeof", "static", "static_assert",
            "static_cast", "struct", "switch", "template", "this", "thread_local", "throw", "true",
            "try", "typedef", "typeid", "typename", "union", "unsigned", "using", "virtual", "void",
            "volatile", "wchar_t", "while", "xor", "xor_eq", "#include", "#define", "#ifdef", "#endif"
        )

        private val swiftTypes = setOf(
            "Int", "String", "Bool", "Double", "Float", "Character", "Array", "Dictionary",
            "Set", "Optional", "Any", "AnyObject", "Void", "Never", "Result", "Error",
            "UIView", "UIViewController", "UILabel", "UIButton", "UIImage", "UIColor",
            "CGRect", "CGPoint", "CGSize", "CGFloat", "NSObject", "NSString", "NSArray"
        )

        private val tsTypes = setOf(
            "Array", "Object", "String", "Number", "Boolean", "Function", "Promise",
            "Map", "Set", "WeakMap", "WeakSet", "Date", "RegExp", "Error", "Symbol",
            "Record", "Partial", "Required", "Readonly", "Pick", "Omit", "Exclude", "Extract",
            "Component", "Observable", "Subject", "EventEmitter"
        )

        // Pre-compiled regex patterns
        private val stringPattern = Pattern.compile("\"[^\"\\\\]*(?:\\\\.[^\"\\\\]*)*\"|'[^'\\\\]*(?:\\\\.[^'\\\\]*)*'|`[^`]*`")
        private val commentPattern = Pattern.compile("//[^\\n]*|/\\*[\\s\\S]*?\\*/|#[^\\n]*")
        private val numberPattern = Pattern.compile("\\b\\d+\\.?\\d*\\b")
        private val wordPattern = Pattern.compile("\\b[a-zA-Z_\$][a-zA-Z0-9_\$]*\\b")
        private val decoratorPattern = Pattern.compile("@[a-zA-Z_][a-zA-Z0-9_]*")
        
        // Decorator color (purple/magenta like VSCode) - #C586C0
        private val decoratorColor = Color.rgb(197, 134, 192)
    }

    private val executors = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())

    /**
     * Highlight code synchronously
     */
    @JvmOverloads
    fun highlight(code: String, language: String, scheme: ColorScheme = _darkScheme): SpannableStringBuilder {
        val spannableString = SpannableStringBuilder(code)

        // Get keywords and types for the language
        val (keywords, types) = when (language.lowercase()) {
            "swift" -> swiftKeywords to swiftTypes
            "typescript", "ts", "javascript", "js" -> tsKeywords to tsTypes
            "python", "py" -> pythonKeywords to emptySet()
            "java" -> javaKeywords to emptySet()
            "kotlin", "kt" -> kotlinKeywords to emptySet()
            "go", "golang" -> goKeywords to emptySet()
            "rust", "rs" -> rustKeywords to emptySet()
            "c", "cpp", "c++", "cxx", "h", "hpp" -> cppKeywords to emptySet()
            else -> tsKeywords to tsTypes
        }

        // Apply monospace font span
        spannableString.setSpan(
            TypefaceSpan("monospace"),
            0,
            code.length,
            Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )

        // Base color
        spannableString.setSpan(
            ForegroundColorSpan(scheme.plain),
            0,
            code.length,
            Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )

        // Track colored ranges to prevent overlap
        val coloredRanges = mutableListOf<IntRange>()

        // 1. Comments first (highest precedence)
        val commentMatcher = commentPattern.matcher(code)
        while (commentMatcher.find()) {
            val range = commentMatcher.start() until commentMatcher.end()
            spannableString.setSpan(
                ForegroundColorSpan(scheme.comment),
                range.first,
                range.last + 1,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            spannableString.setSpan(
                StyleSpan(Typeface.ITALIC),
                range.first,
                range.last + 1,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            coloredRanges.add(range)
        }

        // 2. Strings
        val stringMatcher = stringPattern.matcher(code)
        while (stringMatcher.find()) {
            val range = stringMatcher.start() until stringMatcher.end()
            if (!coloredRanges.any { it.overlaps(range) }) {
                spannableString.setSpan(
                    ForegroundColorSpan(scheme.string),
                    range.first,
                    range.last + 1,
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                )
                coloredRanges.add(range)
            }
        }

        // 3. Decorators/Annotations (@Component, @objc, etc.)
        val decoratorMatcher = decoratorPattern.matcher(code)
        while (decoratorMatcher.find()) {
            val range = decoratorMatcher.start() until decoratorMatcher.end()
            if (!coloredRanges.any { it.overlaps(range) }) {
                spannableString.setSpan(
                    ForegroundColorSpan(decoratorColor),
                    range.first,
                    range.last + 1,
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                )
                coloredRanges.add(range)
            }
        }

        // 4. Numbers
        val numberMatcher = numberPattern.matcher(code)
        while (numberMatcher.find()) {
            val range = numberMatcher.start() until numberMatcher.end()
            if (!coloredRanges.any { it.overlaps(range) }) {
                spannableString.setSpan(
                    ForegroundColorSpan(scheme.number),
                    range.first,
                    range.last + 1,
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                )
            }
        }

        // 5. Keywords and identifiers
        val wordMatcher = wordPattern.matcher(code)
        while (wordMatcher.find()) {
            val range = wordMatcher.start() until wordMatcher.end()
            if (coloredRanges.any { it.overlaps(range) }) continue

            val word = code.substring(range.first, range.last + 1)

            when {
                keywords.contains(word) -> {
                    spannableString.setSpan(
                        ForegroundColorSpan(scheme.keyword),
                        range.first,
                        range.last + 1,
                        Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                    spannableString.setSpan(
                        StyleSpan(Typeface.BOLD),
                        range.first,
                        range.last + 1,
                        Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }
                types.contains(word) || (word.isNotEmpty() && word[0].isUpperCase() && word.length > 1) -> {
                    spannableString.setSpan(
                        ForegroundColorSpan(scheme.type),
                        range.first,
                        range.last + 1,
                        Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }
                else -> {
                    // Check if followed by ( for function
                    val afterWord = range.last + 1
                    if (afterWord < code.length) {
                        var idx = afterWord
                        while (idx < code.length && (code[idx] == ' ' || code[idx] == '\t')) {
                            idx++
                        }
                        if (idx < code.length && code[idx] == '(') {
                            spannableString.setSpan(
                                ForegroundColorSpan(scheme.function),
                                range.first,
                                range.last + 1,
                                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                            )
                        }
                    }
                }
            }
        }

        return spannableString
    }

    /**
     * Full highlight with tokenization (alias for highlight)
     */
    @JvmOverloads
    fun highlightFull(code: String, language: String, scheme: ColorScheme = _darkScheme): SpannableStringBuilder {
        return highlight(code, language, scheme)
    }

    /**
     * Highlight code asynchronously and update a TextView
     */
    @JvmOverloads
    fun highlightAsync(
        code: String,
        language: String,
        textView: TextView,
        scheme: ColorScheme = _darkScheme
    ) {
        executors.execute {
            val highlighted = highlight(code, language, scheme)
            mainHandler.post {
                textView.text = highlighted
            }
        }
    }

    /**
     * Highlight code asynchronously with callback
     */
    interface HighlightCallback {
        fun onComplete(result: SpannableStringBuilder)
    }

    @JvmOverloads
    fun highlightAsync(
        code: String,
        language: String,
        callback: HighlightCallback,
        scheme: ColorScheme = _darkScheme
    ) {
        executors.execute {
            val highlighted = highlight(code, language, scheme)
            mainHandler.post {
                callback.onComplete(highlighted)
            }
        }
    }

    // Extension function to check if ranges overlap
    private fun IntRange.overlaps(other: IntRange): Boolean {
        return this.first <= other.last && other.first <= this.last
    }
}
